File: haskell/glue/src/Glue/Lib/Math/Utility/Min.hs
module Glue.Lib.Math.Utility.Min where

import Glue.Eval (Eval, eval, throwError)
import Glue.Eval.Exception
import Glue.IR (IR (..))

-- Minimum function
-- Mirrors Haskell Glue.Lib.Math.Utility.Min.min exactly
min :: IR Eval
min = NativeFunc minImpl

-- Minimum function implementation
-- Mirrors Haskell Glue.Lib.Math.Utility.Min.minImpl exactly
minImpl :: IR Eval -> Eval (IR Eval)
minImpl arg1 = pure $ NativeFunc (minWith arg1)

minWith :: IR Eval -> IR Eval -> Eval (IR Eval)
minWith arg1 arg2 = do
    va1 <- eval arg1
    va2 <- eval arg2
    case (va1, va2) of
        (Integer n1, Integer n2) -> pure $ Integer (Prelude.min n1 n2)
        (Float n1, Float n2) -> pure $ Float (Prelude.min n1 n2)
        (Integer n1, Float n2) -> pure $ Float (Prelude.min (fromIntegral n1) n2)
        (Float n1, Integer n2) -> pure $ Float (Prelude.min n1 (fromIntegral n2))
        _ -> throwError $ wrongArgumentType ["number", "number"]

File: dart/glue/lib/src/lib/math/utility/min.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Minimum function (returns the smaller of two numbers)
/// Mirrors Haskell Glue.Lib.Math.Utility.Min.min exactly
final Ir min = IrNativeFunc(minImpl);

/// Minimum function implementation (returns the smaller of two numbers)
/// Mirrors Haskell Glue.Lib.Math.Utility.Min.minImpl exactly
Eval<Ir> minImpl(Ir left) {
  return Eval.pure(IrNativeFunc(minWith(left)));
}

/// Helper function for second argument
/// Mirrors Haskell Glue.Lib.Math.Utility.Min.minWith exactly
Eval<Ir> Function(Ir) minWith(Ir left) {
  return (Ir right) => switch ((left, right)) {
    (IrInteger(value: final n1), IrInteger(value: final n2)) => Eval.pure(
      IrInteger(n1 < n2 ? n1 : n2),
    ),
    (IrFloat(value: final n1), IrFloat(value: final n2)) => Eval.pure(
      IrFloat(n1 < n2 ? n1 : n2),
    ),
    (IrInteger(value: final n1), IrFloat(value: final n2)) => Eval.pure(
      IrFloat(n1 < n2 ? n1.toDouble() : n2),
    ),
    (IrFloat(value: final n1), IrInteger(value: final n2)) => Eval.pure(
      IrFloat(n1 < n2 ? n1 : n2.toDouble()),
    ),
    _ => throwError(wrongArgumentType(['number', 'number'])),
  };
}
